### PR TITLE
Bump nmstate to 0.3

### PR DESCRIPTION
**Pull request by qinqon**
**Is this a BUG FIX or a FEATURE ?**:
Looks like nmstate 0.2 is going to approach EOL so we bump to next minor
it also depends on NM 1.26. We are not going to bump kubecirtci node NM
since we have to check that this is retrocompatible.

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind enhancement

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Release note**:

```release-note
Bump nmstate to 0.3
```

### CODE REVIEW

**Review by phoracek — APPROVED** [review]
/lgtm
/approve

**Comment by qinqon** [comment]
force-push: rebased

**Comment by qinqon** [comment]
Configration guide issue
```
desired                                                                         
=======                                                                         
---                                                                             
name: eth1                                                                      
type: ethernet                                                                  
state: down                                                                     
ipv4:                                                                           
                                                                                
        \ dhcp: false                                                           
ipv6: {}                                                                        
                                                                                
current                                                                         
=======                                                                         
---                                                                             
name: eth1                                                                      
type: ethernet                                                                  
state:                                                                          
        down                                                                    
ipv4:                                                                           
  enabled: false                                                                
ipv6:                                                                           
  enabled: false                                                                
lldp:                                                                           
  enabled:                                                                      
        false                                                                   
mac-address: 52:55:00:D1:56:02                                                  
mtu: 1500                                                                       
                                                                                
difference                                                                      
==========                                                                      
---                                                                             
        desired                                                                 
+++ current                                                                     
@@ -3,5 +3,10 @@                                                                
 type: ethernet                                                                  
 state: down                                                                    
 ipv4:                                                                          
-                                                                               
        \ dhcp: false                                                           
-ipv6: {}                                                                       
+  enabled: false                                                               
+ipv6:                                                                          
+  enabled: false                                                               
+lldp:                                                                          
+                           
```

**Comment by qinqon** [comment]
[force-push](https://github.com/nmstate/kubernetes-nmstate/compare/f8c280bad7dadce8f4f76cf9bfedb6791ce4ebac..ff9896367c7c6f73ab73e40a69023a505da237f0): In crease global timeout to 80m and `kubectl wait` at user guide examples to 4m.

**Comment by qinqon** [comment]
Upgrade test failure

```

    Timed out after 480.000s.
    should reach expected status at NNCP 'test-policy', 
     current enactments statuses:
    node01.bond0-eth1-eth2:
      conditions:
      - lastHearbeatTime: "2020-12-01T12:56:35Z"
        lastTransitionTime: "2020-12-01T12:56:35Z"
        message: |-
          error reconciling NodeNetworkConfigurationPolicy at desired state apply: , failed to execute nmstatectl set --no-commit --timeout 480: 'exit status 1' '' '2020-12-01 12:56:35,839 root         WARNING  libnm version 1.26.2 mismatches NetworkManager version 1.22.8
          Traceback (most recent call last):
            File "/usr/bin/nmstatectl", line 11, in <module>
              load_entry_point('nmstate==0.3.6', 'console_scripts', 'nmstatectl')()
            File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 69, in main
              return args.func(args)
            File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 279, in apply
              args.save_to_disk,
            File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 308, in apply_state
              save_to_disk=save_to_disk,
            File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 69, in apply
              net_state = NetState(desired_state, current_state, save_to_disk)
            File "/usr/lib/python3.6/site-packages/libnmstate/net_state.py", line 40, in __init__
              save_to_disk,
            File "/usr/lib/python3.6/site-packages/libnmstate/ifaces/ifaces.py", line 106, in __init__
              self._pre_edit_validation_and_cleanup()
            File "/usr/lib/python3.6/site-packages/libnmstate/ifaces/ifaces.py", line 128, in _pre_edit_validation_and_cleanup
              self._validate_over_booked_slaves()
            File "/usr/lib/python3.6/site-packages/libnmstate/ifaces/ifaces.py", line 421, in _validate_over_booked_slaves
              f"Interface {iface.name} slave {slave_name} is "
          libnmstate.error.NmstateValueError: Interface bond0 slave eth1 is already enslaved by interface br42
          '
        reason: FailedToConfigure
        status: "True"
        type: Failing
      - lastHearbeatTime: "2020-12-01T12:56:35Z"
        lastTransitionTime: "2020-12-01T12:56:35Z"
        reason: FailedToConfigure
        status: "False"
        type: Available
      - lastHearbeatTime: "2020-12-01T12:56:35Z"
    ```

**Comment by qinqon** [comment]
/retest

Prow is aborting the jobs

**Comment by qinqon** [comment]
/retest

hiccup
```
Step 1/9 : FROM registry.access.redhat.com/ubi8/ubi-minimal
Get https://registry.access.redhat.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Makefile:157: recipe for target 'operator' failed
make[1]: *** [operator] Error 1

```

File: test/e2e/handler/node_selector_test.go [hunk 1]
```diff
@@ -25,8 +25,8 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
 		})
 
 		AfterEach(func() {
-			By(fmt.Sprintf("Deleteting linux bridge %s", bridge1))
-			setDesiredStateWithPolicy(bridge1, linuxBrAbsent(bridge1))
+			By(fmt.Sprintf("Deleteting linux bridge %s at all nodes", bridge1))
+			setDesiredStateWithPolicyWithoutNodeSelector(bridge1, linuxBrAbsent(bridge1))
```
Comment (phoracek) [inline test/e2e/handler/node_selector_test.go:30]: Why do we need this? We configured it only on those matching a label. If it is to remove the nested policy without a label, we should create its own AfterEach in its respective context. As a reader of this Context and its It, this cleanup confuses me.

File: test/e2e/handler/main_test.go [hunk 2]
```diff
@@ -50,7 +51,7 @@ var _ = BeforeSuite(func() {
 
 	testenv.Start()
 
-	By("Getting node list from cluster")
+	By("Getting worker node list from cluster")
```
Comment (phoracek) [inline test/e2e/handler/main_test.go:54]: Please elaborate in the commit message why do we need both `nodes` and `allNodes`. It is not clear to me.

File: test/e2e/handler/states.go [hunk 3]
```diff
@@ -178,6 +178,29 @@ func vlanUpWithStaticIP(iface string, ipAddress string) nmstate.State {
 `, iface, ipAddress))
 }
 
+func resetPrimaryAndSecondaryNICS() nmstate.State {
```
Comment (rhrazdil) [inline test/e2e/handler/states.go:181]: I'm not super sure about the conventions, but I think the last s should be small

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 4]
```diff
@@ -296,7 +296,7 @@ func (r *NodeNetworkConfigurationPolicyReconciler) SetupWithManager(mgr ctrl.Man
 
 	allPolicies := handler.ToRequestsFunc(
 		func(handler.MapObject) []reconcile.Request {
-			log := r.Log.WithValues("allPolicies")
+			log := r.Log.WithName("allPolicies")
```
Comment (phoracek) [inline controllers/nodenetworkconfigurationpolicy_controller.go:299]: Why? Please use the commit message to elaborate whether it is just a side fix you made cause we had issue in logging or whether it is related to "bump nmstate to 0.3"

File: test/e2e/handler/upgrade_test.go [hunk 5]
```diff
@@ -23,6 +23,8 @@ var _ = Describe("NodeNetworkConfigurationPolicy upgrade", func() {
 				},
 				Spec: nmstate.NodeNetworkConfigurationPolicySpec{
 					DesiredState: linuxBrUp(bridge1),
+					NodeSelector: map[string]string{"node-role.kubernetes.io/worker": ""},
+					Parallel:     parallelRollout,
```
Comment (phoracek) [inline test/e2e/handler/upgrade_test.go:27]: This is not related to the PR, is it?

**Review by phoracek — CHANGES_REQUESTED** [review]
Looks good.

In general, I don't mind minor fixes being piggy backed on unrelated PRs. What I need in such cases are proper commit messages though. It is not clear to me why are some commits present here and if I should think about them as a part of the nmstate 0.3 fix or just a cleanup.

Please add motivation to each of the commits. One sentence is ok, but it should be enough to tell me whether it is a sweet little cleanup or part of the main PR effort.

File: test/e2e/handler/node_selector_test.go [hunk 6]
```diff
@@ -81,6 +84,8 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
 						Type:   nmstate.NodeNetworkConfigurationEnactmentConditionMatching,
 						Status: corev1.ConditionTrue,
 					}))
+				//TODO: Remove this when webhook retest policy status when node labels are changed
+				time.Sleep(3 * time.Second)
```
Comment (rhrazdil) [inline test/e2e/handler/node_selector_test.go:88]: is this to make sure we don't hit the available state from before adding the label?

**Review by rhrazdil — COMMENTED** [review]
one nit and one question

**Comment by qinqon** [comment]
/retest

flaky at bond test.

File: test/e2e/handler/states.go (hunk 3 above)
Comment (qinqon) [reply test/e2e/handler/states.go:181]: Do you mean `resetPrimaryAndSecondaryNics`

File: test/e2e/handler/node_selector_test.go (hunk 6 above)
Comment (qinqon) [reply test/e2e/handler/node_selector_test.go:88]: You got it, I am implementing the webhook to prevent that, for now we need this Sleep.

File: test/e2e/handler/main_test.go (hunk 2 above)
Comment (qinqon) [reply test/e2e/handler/main_test.go:54]: Done

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 4 above)
Comment (qinqon) [reply controllers/nodenetworkconfigurationpolicy_controller.go:299]: Done

File: test/e2e/handler/upgrade_test.go (hunk 5 above)
Comment (qinqon) [reply test/e2e/handler/upgrade_test.go:27]: Explaineed at the commit messages.

File: test/e2e/handler/node_selector_test.go (hunk 1 above)
Comment (qinqon) [reply test/e2e/handler/node_selector_test.go:30]: Explained at the commit messages.

**Comment by qinqon** [comment]
[force-push](force-push): Adressed @rhrazdil comments about reset NICs name function

File: test/e2e/handler/upgrade_test.go (hunk 5 above)
Comment (phoracek) [reply test/e2e/handler/upgrade_test.go:27]: I don't see this one being explained, am I just missing it?

**Comment by qinqon** [comment]
/retest
bond flakiness

File: test/e2e/handler/node_selector_test.go (hunk 1 above)
Comment (phoracek) [reply test/e2e/handler/node_selector_test.go:30]: I think we should keep this cleanup and create a new one in the nested. It cleans up the nested policy, it is not needed for this level.

File: test/e2e/handler/upgrade_test.go (hunk 5 above)
Comment (qinqon) [reply test/e2e/handler/upgrade_test.go:27]: The test was being applied a master and workers but being cleaned up only at worked also it was not being executed in parallel at parallel lane.

File: test/e2e/handler/node_selector_test.go (hunk 1 above)
Comment (qinqon) [reply test/e2e/handler/node_selector_test.go:30]: It does the cleanup from the stuff populated at `BeforeEach` so this level make sense.

**Review by phoracek — APPROVED** [review]
/lgtm
/approve
